File: LogicAndTrick.WikiCodeParser/Elements/Element.cs
using System;
using System.Collections.Generic;
using System.Linq;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Elements
{
    public abstract class Element
    {
        public List<string> Scopes { get; set; }
        public int Priority { get; set; } = 0;

        public virtual bool InScope(string scope) => string.IsNullOrWhiteSpace(scope) ||
                                                     Scopes.Contains(scope, StringComparer.InvariantCultureIgnoreCase);

        public abstract bool Matches(Lines lines);
        public abstract INode Consume(Parser parser, ParseData data, Lines lines, string scope);
    }
}

File: src/Elements/Element.ts
import { Lines } from '../Lines';
import { INode } from '../Nodes/INode';
import { ParseData } from '../ParseData';
import { Parser } from '../Parser';

export abstract class Element {
    public Scopes : string[];
    public Priority = 0;

    public InScope(scope : string | null) : boolean {
        return scope == null || scope.trim() == '' || this.Scopes.includes(scope);
    }

    public abstract Matches(lines : Lines) : boolean;
    public abstract Consume(parser : Parser, data : ParseData, lines : Lines, scope : string) : INode;
}